### PR TITLE
Add support for ufcs chaining

### PR DESCRIPTION
**Pull request by filipsajdak**
The chaining starts from the function call `fun().gun().bun()` or the method call `o.fun().gun()`.

It integrates chaining to part of the code that rewrites Cpp2 postfix operators to Cpp1 prefix operators. That allows the use of recurrency to go through the entire expression and insert UFCS in parts that need it.

After this change the Cpp2 code (taken from `mixed-postfix-expression-custom-formatting.cpp2`)
```cpp
test: (a:_) -> std::string = {
    return call( a,
        a.b(a.c)*++, "hello", /* polite
                          greeting
                          goes here */ " there",
        a.d.e( a.f*.g()++, // because f is foobar
              a.h.i(),
              a.j(a.k,a.l) )
        );
}
```
will generate
```cpp
[[nodiscard]] auto test(auto const& a) -> std::string{
    return call( a,
        ++*cpp2::assert_not_null(CPP2_UFCS(b, a, a.c)), "hello", /* polite
                          greeting
                          goes here */ " there",
        CPP2_UFCS(e, a.d, ++CPP2_UFCS_0(g, *cpp2::assert_not_null(a.f)), // because f is foobar
              CPP2_UFCS_0(i, a.h),
              CPP2_UFCS(j, a, a.k, a.l)));

}
```

~~The algorithm skips methods with more than one template argument (even when it happens inside chaining - it is done in a way that compiles fine).~~
This change is handling methods with more than one template arguments - already pushed to the PR.

This Cpp2 code:
```cpp
fun0()*.fun1(1)++.fun2<1,2>()++;
```
will generate:
```cpp
++CPP2_UFCS_TEMPLATE_0(fun2, (<1,2>), ++CPP2_UFCS(fun1, *cpp2::assert_not_null(fun0()), 1));
```

This change is not using UFCS for function calls with at least one argument. So, the functionality presented below is missing. 
```
fun() -> fun()
fun(a) -> CPP2_UFCS_0(fun, a)
fun(a, b) -> CPP2_UFCS(fun, a, b)
```

This change replaces: #17, #18

### CODE REVIEW

**Comment by filipsajdak** [comment]
@hsutter I merged the previous two PRs and merged this into your other solution. I didn't update the tests yet. If the change is good from your point of view, I will update everything.

UFCS is not complete after this change. There is a lack of UFCS support for cases like:
```
fun(a) -> CPP2_UFCS_0(fun, a)
fun(a, b) -> CPP2_UFCS(fun, a, b)
```

one of the issues that blocked me at the moment is the code:
```cpp
std::to_string(x as int);
```
generates
```cpp
CPP2_UFCS_0(std::to_string, cpp2::as<int>(x))
```
that, for some reason, fail to compile.

**Comment by filipsajdak** [comment]
@hsutter I have added support for handling methods and functions with more than one template arguments.

Now instead of
```cpp
CPP2_UFCS_TEMPLATE(FUNCNAME,PARAM1,...)
```
UFCS macro for templates looks:
```cpp
CPP2_UFCS_TEMPLATE(FUNCNAME,TEMPARGS,PARAM1,...)
```
That means that calling
```cpp
obj.fun<int,long,double>(1,2);
```
will be generated to:
```cpp
CPP2_UFCS_TEMPLATE(fun, (<int,long,double>), obj, 1, 2);
```

Working prototype: https://godbolt.org/z/nnoEqz814

**Comment by hsutter** [comment]
Thanks! I'm going to start looking at this next.

> It integrates chaining to part of the code that rewrites Cpp2 postfix operators to Cpp2 prefix operators. 

That sounds really promising, it's exactly what I had in mind.

**Comment by hsutter** [comment]
@filipsajdak Thanks again, I have some clarifying questions...

> UFCS is not complete after this change. There is a lack of UFCS support for cases like:
> `fun(a) -> CPP2_UFCS_0(fun, a)`
> `fun(a, b) -> CPP2_UFCS(fun, a, b)`

Those should not be transformed though, right? The current intent was for UFCS to be only for `x.f(...)` memberlike call notation.

Are you proposing to make the `f(x,...)` nonmember call notation also UFCS, so that it could invoke a member function? I'd rather stick with UFCS-ing member call notation only for now, and it will probably make this PR's implementation easier. How does that sound?

P.S.: In the diff view, I loved seeing the current UFCS extra branch at the top of the function show in "bulk deleted" all-red!. It really should be rolled into the general inside-out transformation loop that's the main workhorse in the function, which is what you're doing.

**Comment by filipsajdak** [comment]
It is just a comment. I was trying to clarify what is working and what is not. 

I have seen both approaches proposed.

I have checked https://isocpp.org/files/papers/N4165.pdf and realized that what is currently implemented is what you want (for sure, it supports auto-completion).

I was trying to implement the other style as well. It is possible but the current change supports your proposed style and is fine for me.

If it is not needed e.g. for generic code let's leave it as is.

**Comment by filipsajdak** [comment]
TLDR: it supports only `x.f(...)`

**Comment by hsutter** [comment]
Ack: Great!

**Comment by filipsajdak** [comment]
Rebased to main.

**Comment by hsutter** [comment]
In-progress review ack: Looking good.

I've now completed the first two review steps:
- A first cursory code review: The basic approach and structure seems right.
- Ran regressions as a quick smoke test: No final execution output changed, so that's looking good. But there were some small Cpp1 whitespace output changes.

Next on my list:
- Investigate those whitespace changes: They might be fine, or might just need passing a different `.position()` somewhere.
- Deeper code review: Think through the new code line by line to make sure I grok each step.

File: source/cppfront.cpp [hunk 1]
```diff
@@ -1928,6 +1933,18 @@ class cppfront
...
 
+        if (args) {
+            // if after printing core expression args is defined
+            // it means that the chaining started by function call
+            // we need to print its arguments
+            suffix.emplace_back(")", n.position());
+            for (auto&& e: args.value()) {
+                suffix.push_back(e);
+            }
+            suffix.emplace_back("(", n.position());
```
Comment (hsutter) [inline source/cppfront.cpp:1944]: I think these two lines that print `"("` and `")"` may be what's causing the whitespace changes I noticed in some of the regression tests. Should we extend `args` to store the positions of the left and right parens, and then use those positions here? Just using `n.position()` is using the position of the whole expression and we're past that... the parens may even be on other lines, as they are in `mixed-function-expression-and-std-for-each.cpp2` and `mixed-postfix-expression-custom-formatting.cpp2`... those are two examples where I saw the `)` get moved up/left instead of keeping its original position.

**Review by hsutter — COMMENTED** [review]
Looks pretty good, mostly minor comments including that I think I found the reason the `(` `)` were changing position in some of the regression test cases. I have a suggestion that should resolve that.

File: include/cpp2util.h [hunk 2]
```diff
@@ -536,21 +536,23 @@ class out {
...
 
-#define CPP2_UFCS_TEMPLATE(FUNCNAME,PARAM1,...) \
+#define CPP2_UFCS_REMPARENS(...) __VA_ARGS__
+
+#define CPP2_UFCS_TEMPLATE(FUNCNAME,TEMPARGS,PARAM1,...) \
 [](auto&& obj, auto&& ...params) { \
-    if constexpr (requires{ std::forward<decltype(obj)>(obj).template FUNCNAME(std::forward<decltype(params)>(params)...); }) { \
-        return std::forward<decltype(obj)>(obj).template FUNCNAME(std::forward<decltype(params)>(params)...); \
+    if constexpr (requires{ std::forward<decltype(obj)>(obj).template FUNCNAME CPP2_UFCS_REMPARENS TEMPARGS (std::forward<decltype(params)>(params)...); }) { \
+        return std::forward<decltype(obj)>(obj).template FUNCNAME CPP2_UFCS_REMPARENS TEMPARGS (std::forward<decltype(params)>(params)...); \
```
[the remark this answers is not in the document]
Comment (filipsajdak) [reply include/cpp2util.h:544]: I have learned that here: https://www.scs.stanford.edu/~dm/blog/va-opt.html while working on https://github.com/hsutter/cppfront/pull/47

I am checking the link you have sent and will check the proposed name.

File: source/cppfront.cpp [hunk 3]
```diff
@@ -1828,9 +1776,60 @@ class cppfront
...
+
+                auto ufcs_string = std::string("CPP2_UFCS");
+
+                if (i->id_expr->template_args_count() > 0) {
+                    ufcs_string += "_TEMPLATE";
+                    // we need to replace "fun<int,long,double>" to "fun, (<int,long,double>)" to be able to generate
+                    // from obj.fun<int, long, double>(1,2) this CPP2_UFCS_TEMPLATE(fun, (<int,long, double>), obj, 1, 2)
+                    auto split = funcname.find('<'); assert(split != std::string::npos);
+                    funcname.insert(split, ", (");
+                    funcname += ')';
```
[the remark this answers is not in the document]
Comment (filipsajdak) [reply source/cppfront.cpp:1812]: At this point we know that it is a template so it should have `<` and `>` - assertion is the right thing to do here.

I will add it.

File: source/cppfront.cpp [hunk 4]
```diff
@@ -1874,21 +1873,25 @@ class cppfront
...
 
                 if (i->id_expr) {
-                    auto print = std::string{};
-                    printer.emit_to_string(&print);
-                    emit(*i->id_expr, false /*not a local name*/);
-                    printer.emit_to_string();
+
+                    if (args) {
+                        // if args are stored it means that this is function or method
+                        // that is not handled by UFCS e.g. that has more than one template argument
```
[the remark this answers is not in the document]
Comment (filipsajdak) [reply source/cppfront.cpp:1879]: That comment was for you to emphasise that it will work also for more template arguments.

File: source/cppfront.cpp [hunk 5]
```diff
@@ -1805,6 +1734,25 @@ class cppfront
...
 
+        auto args = std::optional<std::vector<text_with_pos>>{};
+
+        auto print_to_string = [&](auto& i, auto... args) {
+            auto print = std::string{};
+            printer.emit_to_string(&print);
+            emit(i, args...);
+            printer.emit_to_string();
+            return print;
+        };
```
[the remark this answers is not in the document]
Comment (filipsajdak) [reply source/cppfront.cpp:1745]: I am glad that you like it.

File: source/cppfront.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (filipsajdak) [reply source/cppfront.cpp:1879]: Ah... Now I see that I did not update the comment - will correct that.

File: source/cppfront.cpp (hunk 1 above)
Comment (filipsajdak) [reply source/cppfront.cpp:1944]: Yes, you are right. I struggled to find a solution as I know it is essential for you.

I tried to use `push_need_expression_list_parens(true)` at the generation of args, but something was wrong. I will test it once again.

File: include/cpp2util.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (filipsajdak) [reply include/cpp2util.h:544]: I don't like the `CPP2_VA_ARGS_IF_PARENS` name. `COMMA_IF_PARENS` is used when there are parentheses, then it will expand to the comma. In contrast, `CPP2_VA_ARGS_IF_PARENS` suggests that it will expand to the `__VA_ARGS__`, which is misleading. In the context of our macro, we use `__VA_ARGS__` but from the perspective of the caller it is just removing parentheses or there is also used the name `EXPAND` that might replace `REMPARENS`.

From my perspective, `REMPARENS` is more descriptive.

File: source/cppfront.cpp (hunk 1 above)
Comment (hsutter) [reply source/cppfront.cpp:1944]: You shouldn't have to push the parens state... The basic idea would be to store the open and close paren positions alongside `args` itself, either as `source_position`s or as `token const*`s to the paren tokens so we can ask them their positions.

But I can also go do this post-merge if that's the only thing holding up the merge, don't let this alone block the PR.

Comment (hsutter) [reply source/cppfront.cpp:1944]: OK, I've found and tested the solution to correct the paren positions. I'll post it here as another review, it's just two more lines to remember the `(` and `)` position and then using them in the three places.

File: source/cppfront.cpp [hunk 6]
```diff
@@ -1805,6 +1734,25 @@ class cppfront
         auto last_was_prefixed = false;
         auto saw_dollar        = false;
 
+        auto args = std::optional<std::vector<text_with_pos>>{};
```
Comment (hsutter) [inline source/cppfront.cpp:1737]: Here, store two more values (we could bundle these three in a struct but I'm fine with three similarly named locals):

```
        auto args_open_pos  = source_position{};
        auto args_close_pos = source_position{};
```

File: source/cppfront.cpp [hunk 7]
```diff
@@ -1828,9 +1776,60 @@ class cppfront
                 }
             }
 
+            // Going backwards if we found LeftParen it might be UFCS
+            // expr_list is emited to args variable for future use
+            if (i->op->type() == lexeme::LeftParen) {
+
```
Comment (hsutter) [inline source/cppfront.cpp:1782]: Here, add:

```
                assert(i->op_close);
                args_open_pos  = i->op->position();
                args_close_pos = i->op_close->position();
```

File: source/cppfront.cpp [hunk 8]
```diff
@@ -1828,9 +1776,60 @@ class cppfront
...
+                    funcname.insert(split, ", (");
+                    funcname += ')';
+                }
+                //  If there are no additional arguments, use the _0 version
+                if (args.value().empty()) {
+                    ufcs_string += "_0";
+                }
+
+                prefix.emplace_back(ufcs_string + "(" + funcname + ", ", i->op->position() );
+                suffix.emplace_back(")", i->op->position() );
```
Comment (hsutter) [inline source/cppfront.cpp:1820]: Change these two lines to:

```
                prefix.emplace_back(ufcs_string + "(" + funcname + ", ", args_open_pos );
                suffix.emplace_back(")", args_close_pos );
```

File: source/cppfront.cpp [hunk 9]
```diff
@@ -1828,9 +1776,60 @@ class cppfront
...
+                if (!args.value().empty()) {
+                    for (auto&& e: args.value()) {
+                        suffix.push_back(e);
+                    }
+                    suffix.emplace_back(", ", i->op->position());
```
Comment (hsutter) [inline source/cppfront.cpp:1825]: (Aside: This position isn't precise but is probably fine because we're rewriting so much, and this position is probably earlier than we're currently at which means it won't move anything. There are other places, particularly in this function, where I don't use a precise position for similar reasons.)

File: source/cppfront.cpp (hunk 1 above)
Comment (hsutter) [inline source/cppfront.cpp:1944]: These two `suffix.` lines should also use the positions:

```
            suffix.emplace_back(")", args_close_pos);
            ...
            suffix.emplace_back("(", args_open_pos);
```

That should do it!

File: source/cppfront.cpp [hunk 10]
```diff
@@ -1874,21 +1873,25 @@ class cppfront
...
+                        // that is not handled by UFCS e.g. that has more than one template argument
+                        suffix.emplace_back(")", n.position());
+                        for (auto&& e: args.value()) {
+                            suffix.push_back(e);
+                        }
+                        suffix.emplace_back("(", n.position());
+                        args.reset();
+                    }
+
+                    auto print = print_to_string(*i->id_expr, false /*not a local name*/);
```
Comment (hsutter) [inline source/cppfront.cpp:1888]: These two `suffix.` lines should also use the positions:

```
                        suffix.emplace_back(")", args_close_pos);
                        ...
                        suffix.emplace_back("(", args_open_pos);
```

**Review by hsutter — COMMENTED** [review]
This review shows the fix for preserving the `(` `)` positions.

File: source/cppfront.cpp (hunk 6 above)
Comment (filipsajdak) [reply source/cppfront.cpp:1737]: That was also one of my ideas that I will try.

File: source/cppfront.cpp (hunk 10 above)
Comment (filipsajdak) [reply source/cppfront.cpp:1888]: Ok. Thank you for all your comments. I will introduce these changes.

**Comment by filipsajdak** [comment]
@hsutter I have sent an updated code.

**Comment by hsutter** [comment]
Looks good, merging... thanks, this is a big improvement.

**Comment by filipsajdak** [comment]
Thank you. If you see any other area I might help with, please let me know.

I will do my best to help.
